File: merge/metadata/metadata.go
package metadata

import (
	"context"
	"database/sql"
	"encoding/json"
	"fmt"
	"github.com/gigapi/gigapi/v2/utils/tempconf"
	"github.com/lib/pq"
	_ "github.com/lib/pq"
	"log"
	"strconv"
)

var mdb *sql.DB

func getMetadataDB() (*sql.DB, error) {
	if mdb != nil {
		return mdb, nil
	}
	connStr := tempconf.GetPostgresConnectionString()
	if connStr == "" {
		log.Fatal("POSTGRES environment variables is not set")
	}

	var err error
	mdb, err = sql.Open("postgres", connStr)
	if err != nil {
		return nil, err
	}

	// Test the connection
	err = mdb.Ping()
	if err != nil {
		return nil, err
	}

	return mdb, nil
}

type TableDesc struct {
	Id      int
	Name    string
	Path    string
	OrderBy []string
}

type ColumnDesc struct {
	Id           int64  `json:"column_id"`
	Name         string `json:"column_name"`
	Min          string `json:"min_value"`
	Max          string `json:"max_value"`
	Count        int64  `json:"value_count"`
	NullCount    int64  `json:"null_count"`
	ContainsNans bool   `json:"contains_nans"`
	SizeBytes    int64  `json:"column_size_bytes"`
}

type FilePartitionValue struct {
	PartitionKeyIndex int64  `json:"partition_key_index"`
	PartitionValue    string `json:"partition_value"`
}

type FileDesc struct {
	Id                  int64
	Table               TableDesc
	Path                string
	SizeBytes           int64
	FooterSizeBytes     int64
	RecordCount         int64
	ColumnStats         []ColumnDesc
	FilePartitionValues []FilePartitionValue
	PartitionId         int64
}

func GetFiles(table string, iteration int, withStats bool) ([]FileDesc, error) {
	db, err := getMetadataDB()
	if err != nil {
		return nil, err
	}
	req, args, err := (&fileSeeker{
		Iteration: iteration,
		Table:     table,
		WithStats: withStats,
	}).Build()
	rows, err := db.Query(req, args...)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	var files []FileDesc
	for rows.Next() {
		var (
			columnIds           []int64
			columnSizes         []int64
			valueCounts         []int64
			nullCounts          []int64
			minValues           []string
			maxValues           []string
			containsNans        []bool
			columnNames         []string
			partitionKeyIndices []int64
			partitionValues     []string
		)
		var file FileDesc
		err := rows.Scan(&file.Table.Id, &file.Table.Name, &file.Table.Path, &file.Id, &file.Path,
			&file.SizeBytes, &file.FooterSizeBytes, &file.RecordCount,
			pq.Array(&columnIds),
			pq.Array(&columnSizes),
			pq.Array(&valueCounts),
			pq.Array(&nullCounts),
			pq.Array(&minValues),
			pq.Array(&maxValues),
			pq.Array(&containsNans),
			pq.Array(&columnNames),
			pq.Array(&partitionKeyIndices),
			pq.Array(&partitionValues),
			pq.Array(&file.Table.OrderBy),
		)
		if err != nil {
			return nil, err
		}
		for i, id := range columnIds {
			file.ColumnStats = append(file.ColumnStats, ColumnDesc{
				Id:           id,
				Name:         columnNames[i],
				Min:          minValues[i],
				Max:          maxValues[i],
				Count:        valueCounts[i],
				NullCount:    nullCounts[i],
				ContainsNans: containsNans[i],
				SizeBytes:    columnSizes[i],
			})
		}
		for i, id := range partitionKeyIndices {
			file.FilePartitionValues = append(file.FilePartitionValues, FilePartitionValue{
				PartitionKeyIndex: id,
				PartitionValue:    partitionValues[i],
			})
		}
		files = append(files, file)
	}
	return files, nil
}

const (
	//TODO: configuration
	SCHEMA_NAME = "main"
)

func FinishMerge(ctx context.Context, delete []FileDesc, add []FileDesc, table *TableDesc) error {
	db, err := getMetadataDB()
	if err != nil {
		return err
	}

	tx, err := db.BeginTx(ctx, nil)
	if err != nil {
		return err
	}
	var delFileIds []int64
	for _, file := range delete {
		delFileIds = append(delFileIds, file.Id)
	}

	requests := []string{
		`INSERT INTO ducklake_files_scheduled_for_deletion 
SELECT data_file_id, $1 || '/' || $2 || path as path, true as path_is_relative, NOW() as schedule_start
FROM ducklake_data_file
WHERE data_file_id = ANY($3::INT8[])`,
		`WITH files AS (
	select sum(record_count) as rc, sum(file_size_bytes) as fsb FROM ducklake_data_file WHERE data_file_id = ANY($1::INT8[])
) UPDATE ducklake_table_stats SET 
	record_count = record_count - (select rc FROM files), 
	file_size_bytes = file_size_bytes - (SELECT fsb FROM files)
WHERE table_id = $2`,
		`DELETE FROM ducklake_data_file WHERE data_file_id = ANY($1::INT8[])`,
	}
	args := [][]any{
		{SCHEMA_NAME, table.Name, pq.Array(delFileIds)},
		{pq.Array(delFileIds), table.Id},
		{pq.Array(delFileIds)},
	}

	for i := range add {
		columnDesc, _ := json.Marshal(&add[i].ColumnStats)
		fmt.Println(string(columnDesc))
		req := `
WITH a as (
    INSERT INTO ducklake_snapshot
    SELECT max(snapshot_id) + 1, NOW(), max(schema_version), max(next_catalog_id), max(next_file_id) + 1
    FROM (SELECT * FROM ducklake_snapshot ORDER BY snapshot_id DESC LIMIT 1) AS base
    RETURNING *),
b as (INSERT INTO ducklake_snapshot_changes SELECT max(snapshot_id), 'inserted_into_table:' || $1 FROM a RETURNING *),
c as (UPDATE ducklake_table_stats
    SET next_row_id = next_row_id + $2, record_count = record_count + $2, file_size_bytes = file_size_bytes + $3
    WHERE table_id = $4
    RETURNING *),
d as (INSERT INTO ducklake_data_file
    SELECT max(next_file_id) - 1,
           $4,
           max(snapshot_id),
           null,
           null,
           $5,
           true,
           'parquet',
           $2,
           $3,
           $6,
           (select max(next_row_id - $2) FROM c),
           $7 FROM a RETURNING *),
e as (INSERT INTO ducklake_file_column_statistics SELECT
        (select max(next_file_id) - 1 from a) as data_file_id,
        $4 as table_id,
        (elem->>'column_id')::int8 as column_id,
        (elem->>'column_size_bytes')::int8 as column_size_bytes,
        (elem->>'value_count')::int8 as value_count,
        (elem->>'null_count')::int8 as null_count,
        elem->>'min_value' AS min_value,
        elem->>'max_value' AS max_value,
        null as contains_nan
    FROM
        json_array_elements($8::json) AS elem
    RETURNING *)
select * from e;`
		requests = append(requests, req)
		args = append(args, []any{
			strconv.FormatInt(int64(table.Id), 10),
			add[i].RecordCount,
			add[i].SizeBytes,
			table.Id,
			add[i].Path,
			add[i].FooterSizeBytes,
			add[i].PartitionId,
			string(columnDesc),
		})
	}

	for i, r := range requests {
		_, err = tx.Exec(r, args[i]...)
		if err != nil {
			tx.Rollback()
			return err
		}
	}
	return tx.Commit()
}

File: services/metadata/ducklake/metadata.py
import json
import logging
from typing import List, Optional
import asyncpg
from pydantic import BaseModel

from config import settings, postgres_connection_dict
from services.metadata.ducklake.file_seeker import FileSeeker

mdb = None

async def get_metadata_db():

    conn_dict = postgres_connection_dict()
    mdb = await asyncpg.connect(**conn_dict)
    return mdb

class TableDesc(BaseModel):
    id: int
    name: str
    path: str
    order_by: List[str]

class ColumnDesc(BaseModel):
    column_id: int = 0
    column_name: str = ""
    min_value: str = ""
    max_value: str = ""
    value_count: int
    null_count: int
    contains_nans: bool
    column_size_bytes: int

class FilePartitionValue(BaseModel):
    partition_key_index: int
    partition_value: str

class FileDesc(BaseModel):
    id: int
    table: TableDesc
    path: str
    size_bytes: int
    footer_size_bytes: int
    record_count: int
    column_stats: Optional[List[ColumnDesc]]
    file_partition_values: Optional[List[FilePartitionValue]]
    partition_id: int

async def get_files(table: str, iteration: int, with_stats: bool) -> List[FileDesc]:
    db = await get_metadata_db()
    seeker = FileSeeker()
    seeker.table = table
    seeker.iteration = iteration
    seeker.with_stats = with_stats
    query, args = seeker.build()

    rows = await db.fetch(query, *args)

    files = []
    for row in rows:
        ic(row)
        file = FileDesc(
            id=row['file_id'],
            table=TableDesc(
                id=row['table_id'],
                name=row['table_name'],
                path=row['table_path'],
                order_by=row['sort']
            ),
            path=row['file_path'],
            size_bytes=row['file_size_bytes'],
            footer_size_bytes=row['footer_size'],
            record_count=row['record_count'],
            column_stats=[
                ColumnDesc(
                    column_id=col_id,
                    column_name=col_name,
                    min_value=min_val,
                    max_value=max_val,
                    value_count=val_count,
                    null_count=null_count,
                    contains_nans=contains_nan,
                    column_size_bytes=col_size
                )
                for col_id, col_name, min_val, max_val, val_count, null_count, contains_nan, col_size in zip(
                    row['column_ids'], row['column_names'], row['min_values'], row['max_values'],
                    row['value_counts'], row['null_counts'], row['contains_nans'], row['column_sizes']
                )
            ],
            file_partition_values=[
                FilePartitionValue(
                    partition_key_index=key_index,
                    partition_value=value
                )
                for key_index, value in zip(row['partition_key_indices'], row['partition_values'])
            ],
            partition_id=0  # Assuming this is not provided in the query result
        )
        files.append(file)

    return files

SCHEMA_NAME = "main"  # TODO: configuration

async def finish_merge(delete: List[FileDesc], add: List[FileDesc], table: TableDesc):
    db = await get_metadata_db()

    async with db.transaction():
        del_file_ids = [file.id for file in delete]

        # Requests equivalent to the Go version
        await db.execute(
            """
            INSERT INTO ducklake_files_scheduled_for_deletion 
            SELECT data_file_id, $1 || '/' || $2 || path as path, true as path_is_relative, NOW() as schedule_start
            FROM ducklake_data_file
            WHERE data_file_id = ANY($3::INT8[])
            """,
            SCHEMA_NAME, table.name, del_file_ids
        )

        await db.execute(
            """
            WITH files AS (
                SELECT sum(record_count) as rc, sum(file_size_bytes) as fsb 
                FROM ducklake_data_file 
                WHERE data_file_id = ANY($1::INT8[])
            ) 
            UPDATE ducklake_table_stats SET 
                record_count = record_count - (SELECT rc FROM files), 
                file_size_bytes = file_size_bytes - (SELECT fsb FROM files)
            WHERE table_id = $2
            """,
            del_file_ids, table.id
        )

        await db.execute(
            "DELETE FROM ducklake_data_file WHERE data_file_id = ANY($1::INT8[])",
            del_file_ids
        )

        for file in add:
            column_desc = json.dumps([col.dict() for col in file.column_stats])
            ic(file)
            ic(file.file_partition_values)
            file_partition_values = json.dumps([partition_value.__dict__ for partition_value in file.file_partition_values])
            await db.execute(
                """
                WITH a AS (
                    INSERT INTO ducklake_snapshot
                    SELECT max(snapshot_id) + 1, NOW(), max(schema_version), max(next_catalog_id), max(next_file_id) + 1
                    FROM (SELECT * FROM ducklake_snapshot ORDER BY snapshot_id DESC LIMIT 1) AS base
                    RETURNING *
                ),
                b AS (
                    INSERT INTO ducklake_snapshot_changes 
                    SELECT max(snapshot_id), 'inserted_into_table:' || $1 FROM a 
                    RETURNING *
                ),
                c AS (
                    UPDATE ducklake_table_stats
                    SET next_row_id = next_row_id + $2, record_count = record_count + $2, file_size_bytes = file_size_bytes + $3
                    WHERE table_id = $4
                    RETURNING *
                ),
                d AS (
                    INSERT INTO ducklake_data_file
                    SELECT max(next_file_id) - 1,
                           $4,
                           max(snapshot_id),
                           null,
                           null,
                           $5,
                           true,
                           'parquet',
                           $2,
                           $3,
                           $6,
                           (SELECT max(next_row_id - $2) FROM c),
                           $7 FROM a 
                    RETURNING *
                ),
                e AS (
                    INSERT INTO ducklake_file_column_statistics 
                    SELECT
                        (SELECT max(next_file_id) - 1 FROM a) as data_file_id,
                        $4 as table_id,
                        (elem->>'column_id')::int8 as column_id,
                        (elem->>'column_size_bytes')::int8 as column_size_bytes,
                        (elem->>'value_count')::int8 as value_count,
                        (elem->>'null_count')::int8 as null_count,
                        elem->>'min_value' AS min_value,
                        elem->>'max_value' AS max_value,
                        null as contains_nan
                    FROM
                        json_array_elements($8::json) AS elem
                    RETURNING *
                ),
                f AS (
                    INSERT INTO ducklake_file_partition_value
                    SELECT 
                        (SELECT max(next_file_id) - 1 FROM a) as data_file_id,
                        $4 as table_id,
                        (elem->>'partition_key_index')::int8 as partition_key_index,
                        (elem->>'partition_value')::text as partition_value
                    FROM json_array_elements($9::json) AS elem
                    RETURNING *
                )
                SELECT * FROM f;
                """,
                str(table.id),
                file.record_count,
                file.size_bytes,
                table.id,
                file.path,
                file.footer_size_bytes,
                file.partition_id,
                column_desc,
                file_partition_values #TODO
            )